sel/src/Sel/SecretKey/AuthenticatedEncryption.hs

```diff
@@ -0,0 +1,208 @@
+-- >   encryptedMessage <- AuthenticatedEncryption.encrypt "hello hello" secretKey nonce
+-- >   let result = AuthenticatedEncryption.decrypt encryptedMessage secretKey nonce
+-- >   print result
+-- >   -- "Just \"hello hello\""
+
+-- | A secret key of size 'cryptoSecretboxKeyBytes'.
+--
+-- @since 0.0.1.0
+newtype SecretKey = SecretKey (ForeignPtr CUChar)
+
+-- |
+--
+-- @since 0.0.1.0
+instance Eq SecretKey where
+  (SecretKey hk1) == (SecretKey hk2) =
+    unsafeDupablePerformIO $
+      foreignPtrEq hk1 hk2 cryptoSecretboxKeyBytes
+
+-- |
+--
+-- @since 0.0.1.0
+instance Ord SecretKey where
+  compare (SecretKey hk1) (SecretKey hk2) =
+    unsafeDupablePerformIO $
+      foreignPtrOrd hk1 hk2 cryptoSecretboxKeyBytes
+
+-- | A random number that must only be used once per exchanged message.
+-- It does not have to be confidential.
+-- It is of size 'cryptoSecretboxNonceBytes'.
+--
+-- @since 0.0.1.0
+newtype Nonce = Nonce (ForeignPtr CUChar)
+
+-- |
+--
+-- @since 0.0.1.0
+instance Eq Nonce where
+  (Nonce hk1) == (Nonce hk2) =
+    unsafeDupablePerformIO $
+      foreignPtrEq hk1 hk2 cryptoSecretboxKeyBytes
+
+-- |
+--
+-- @since 0.0.1.0
+instance Ord Nonce where
+  compare (Nonce hk1) (Nonce hk2) =
+    unsafeDupablePerformIO $
+      foreignPtrOrd hk1 hk2 cryptoSecretboxKeyBytes
+
+-- | Generate a new random secret key.
+--
+-- @since 0.0.1.0
+newSecretKey :: IO SecretKey
+newSecretKey = do
+  fPtr <- Foreign.mallocForeignPtrBytes (fromIntegral cryptoSecretboxKeyBytes)
+  Foreign.withForeignPtr fPtr $ \ptr ->
+    cryptoSecretboxKeygen ptr
+  pure $ SecretKey fPtr
+
+-- | Generate a new random nonce.
+-- Only use it once per exchanged message.
+--
+-- @since 0.0.1.0
+newNonce :: IO Nonce
+newNonce = do
+  (fPtr :: ForeignPtr CUChar) <- Foreign.mallocForeignPtrBytes (fromIntegral cryptoSecretboxNonceBytes)
+  Foreign.withForeignPtr fPtr $ \ptr ->
+    randombytesBuf (Foreign.castPtr @CUChar @Word8 ptr) cryptoSecretboxNonceBytes
+  pure $ Nonce fPtr
+
+-- | A ciphertext consisting of an encrypted message and an authentication tag.
+--
+-- @since 0.0.1.0
+data Hash = Hash
+  { messageLength :: CULLong
+  , hashForeignPtr :: ForeignPtr CUChar
+  }
+
+-- |
+--
+-- @since 0.0.1.0
+instance Eq Hash where
+  (Hash messageLength1 hk1) == (Hash messageLength2 hk2) =
+    unsafeDupablePerformIO $ do
+      result1 <- foreignPtrEq hk1 hk2 (fromIntegral messageLength1 + cryptoSecretboxMACBytes)
+      pure $ (messageLength1 == messageLength2) && result1
+
+-- |
+--
+-- @since 0.0.1.0
+instance Ord Hash where
+  compare (Hash messageLength1 hk1) (Hash messageLength2 hk2) =
+    unsafeDupablePerformIO $ do
+      result1 <- foreignPtrOrd hk1 hk2 (fromIntegral messageLength1 + cryptoSecretboxMACBytes)
+      pure $ compare messageLength1 messageLength2 <> result1
+
+-- | Create an authenticated hash from a message, a secret key that must remain secret, and a one-time cryptographic nonce
```

Comment:
"A secret key that must remain secret" is redundant. The secret key is already secret.